### ISSUE TITLE
invalid column name 'id'

### ISSUE THREAD

**Issue by williambuchanan2** [DuendeSoftware/Support#318]
**Which version of Duende IdentityServer are you using?**

**Which version of .NET are you using?**
6.0

**Describe the bug**

I just upgraded the following packages in Visual Studio nugget: 

Duende.IdentityServer.AspNetIdentity Version="6.1.7" 
Duende.IdentityServer.EntityFramework Version="6.1.7" 

I was on v6.0.4 before. 

Since then I get this error: 

invalid column name 'id' 

It seems to be looking for this column in the PersistedGrants table. 

I tried deleting all the duende tables - assuming it would regenerate them properly. This doesn't seem to have happened. Instead I now get this error: 

'Invalid column name 'CoordinateLifetimeWithUserSession' 

So, it seems to have regenerated the tables, but seems to have forgotten to create some columns. 

**To Reproduce**

Steps to reproduce the behavior.

**Expected behavior**

**Log output/exception with stacktrace**

```
data
```

**Additional context**

**Comment by josephdecock** [comment]
Between versions 6.0 and 6.1, there were some schema changes, which are documented [here](https://docs.duendesoftware.com/identityserver/v6/upgrades/v6.0_to_v6.1/). Also on that page there are the commands needed to run EF migrations to update a 6.0 database to 6.1, if that's the way you wish to manage your schema changes. 

IdentityServer's data stores are designed with flexibility in mind. We don't put many requirements on how your operational and configuration data is stored - as long as you can fulfill the [data store interfaces](https://docs.duendesoftware.com/identityserver/v6/reference/stores/), that's the abstraction that IdentityServer uses. We do provide EntityFramework implementations of those interfaces, but again we are aiming for flexibility. We don't assume anything about how your organization manages database changes, or even which database you're using. Instead, we provide the entity classes that our EF implementation uses, and we allow you to apply schema changes with whatever tools you like. And if you are using EF migrations, we have guidance on how to generate them in our release notes.

**Comment by williambuchanan2** [comment]
Hi Joseph

I appreciate the reply, but i'm now more confused than I was.

I think the documentation is lacking in detail - or I possibly missed something. Either way there doesn't seem to be enough detail to work out what I am supposed to do to get this working. The bunch of links to what looks like support tickets - not sure what they are telling me; and the commands on the page just give me an error ("dotnet : Could not execute because the specified command or file was not found"). I tried a package manager console and VS command window. Maybe some information on where I need to run the commands, or full instructions would be helpful.

So, we are using Duende.IdentityServer.EntityFramework, and it seems this generates the schema for us automatically when the software runs - but then it looks like it doesn't generate the full schema properly, so it seems we need to work out what schema changes are needed to the duende tables to make it work. Could it not just generate the right schema in the first place? 

I guess I just need someone to tell me using clear instructions what I need to do to get this working again...

**Comment by josephdecock** [comment]
It sounds like you're running the entity framework migrations in the wrong directory, or perhaps have your solution organized differently. You probably need to adjust some paths in the command.

In the spirit of "teaching a man to fish", can I recommend that you take a look at our series of Quickstart documentation series? In particular, [quickstart 4](https://docs.duendesoftware.com/identityserver/v6/quickstarts/4_ef/) shows how to set up entity framework and uses the migrations tool. It goes step by step and says exactly where to put each component and where to run the command line commands, so hopefully you'll see how things fit together.

**Comment by williambuchanan2** [comment]
Hi Joseph

Thanks for the reply. I like the idea of teaching a man to fish, but I think my initial complaint about the poor documentation still stands even when trying to follow the quick start guides (something that our original developer complained about as well). 

The problem is the instructions aren't easy to follow. In fact in a lot of cases the instructions just seem like random snippets of information which are out of context and have no bearing on what you are actually trying to do. There are steps missing and lots of things which are not explained. E.g. what is the 'IdentityServer' directory? I'm sure you know exactly what directory that is but I don't, and can't see any instructions telling me where it is. How do I find it (note: we don't have a folder named IdentityServer in our solution). Is it the case that we just need to spend our valuable time working it all out for ourselves with just the bear minimum instruction? My feeling is the instructions have been written by a developer rather than a user.

There is also nothing about what happens when you get an error  - which is exactly what happens when I run pretty much every command listed in the instructions.

Our original developer nearly gave up implementing this, and now I can understand why. I don't understand why a simple package upgrade is so difficult to the point where everyone who attempts to do it needs to learn the whole duende suite from the ground up to achieve it. Why can't I just do an update in Nuget and get on with my life?

Bottom line - i'm still no further forward, and quickly thinking of giving up...

**Comment by williambuchanan2** [comment]
Ok, progress... I worked out what the 'IdentityServer' directory is. I also worked out that I need to run the commands not in the IdentityServer directory but in the main application directory. I then discovered that I need to run the commands with the addition of the --project switch which needs to point back to the IdentityServer directory.

None of the above is documented in the EF upgrade documentation. If it had been I would have saved days of lost time...

**Comment by josephdecock** [comment]
I'm sorry that this has been so frustrating for you. I really care about our docs and want to make them better, and I want your experience to be better too.

The quickstart documentation series walks you through creating an identity server solution and shows how to do various things with it. In the first quickstart, we walk through the creation of the project structure. Each of the following quickstarts build on that. As we note at the start of each quickstart, you can do them all in order, or jump in in the middle by downloading the previous quickstart's code. In the 4th quickstart (that I linked above) the IdentityServer directory refers to the path src/IdentityServer within that example solution. I've added a todo item to our docs to make sure all our instructions in the quickstarts to run command line commands specify the full path to run them in. That makes sense in the quickstart, because it is a tutorial that guides you through creating your solution and therefore expects that your solution is laid out the way it asks you to. 

We can't do the same thing in our upgrade docs, unfortunately, because outside tutorials our users organize their projects within their solutions in all kinds of different ways. IdentityServer doesn't force you to have any particular structure for your solution, so our upgrade docs only give an example of how to use EF. We're expecting (and do note this in our upgrade docs) that you'll modify the paths as appropriate for your solution.

I'm glad that you've made progress. Is there anything else you need help with?

**Comment by williambuchanan2** [comment]
I understand what you are saying but disagree that the quickstart tutorials are clear, and also disagree that the upgrade steps are in any way helpful.

The main point I am making is that once this is implemented (something that 1 developer does once) nobody else really needs to understand the whole duende system. We shouldn't be forced to make everyone who wants to do a simple nuget update go through the entire training program just to understand how to upgrade. In the end it was quite simple - in fact I have now written a batch file to do it for us so we don't have this trouble in future. 

Why not just have a very clear step by step guide for the upgrade path (including potential errors you may encounter, and some troubleshooting) - that way the 99% of people just wanting to upgrade won't need to spend a week learning duende? 

As for the 'IdentityServer' folder - this is key information to know. Without knowing what this means you are blind, so you need to be able to describe this to people so they understand where to run these commands. I get that everyone will have named it differently but we just need to know how to identify this folder.

To make up for my moaning i'm even happy to work with you on the documentation - i'll give you an idiots perspective :-)

**Comment by josephdecock** [comment]
Thank you, I'd love to hear more of your thoughts on where the docs weren't clear to you, especially where there were missing steps in the quickstarts.

**Comment by brockallen** [comment]
Hi @williambuchanan2 -- anything else to add here?

**Comment by williambuchanan2** [comment]
No i'm good now. Thanks
